python/hail/methods/statgen.py

```diff
@@ -2025,9 +2025,170 @@ def annotate_entries(self, **named_exprs):
+        self.annotate_entries(
+            GT = functions.call(functions.gt_from_pl(newPL)),
+            AD = functions.cond(
+                functions.is_defined(ds.AD),
+                functions.range(0, self.new_v.num_alleles()).map(
+                    lambda newi: ds.AD[self.new_to_old[newi]]),
+                functions.null(TArray(TInt32()))),
+            # DP unchanged
+            GQ = functions.gq_from_pl(newPL),
+            PL = newPL)
+
+    def downcode_entries_hts(self):
+        """Use the downcode algorithm to update the matrix table entries for
+        the new, allele filtered variant.  This requires the entries
+        have the HTS genotype schema:
+
+        **The downcode algorithm**
+
+        .. code-block:: text
+
+            GT: Call,
+            AD: Array[!Int32],
+            DP: Int32,
+            GQ: Int32,
+            PL: Array[!Int32].
+
+        The downcode algorithm recodes occurances of filtered alleles
+        to occurances of the reference allele (e.g. 1 -> 0 in our
+        example). So the depths of filtered alleles in the AD field
+        are added to the depth of the reference allele. Where
+        downcodeing filtered alleles merges distinct genotypes, the
+        minimum PL is used (since PL is on a log scale, this roughly
+        corresponds to adding probabilities). The PLs are then
+        re-normalized (shifted) so that the most likely genotype has a
+        PL of 0, and GT is set to this genotype.  If an allele is
+        filtered, this algorithm acts similarly to
+        :meth:`~hail.VariantDataset.split_multi`.
+
+        The downcoding algorithm would produce the following:
+
+        .. code-block:: text
+
+          GT: 0/1
+          GQ: 10
+          AD: 35,50
+
+          0 | 20
+          1 | 0    10
+            +-----------
+              0    1
+
+        In summary:
+
+        - GT: Downcode filtered alleles to reference.
+        - AD: The filtered alleles' columns are eliminated and their
+          value is added to the reference, e.g., filtering alleles 1
+          and 2 transforms ``25,5,10,20`` to ``40,20``.
+        - DP: No change.
+        - PL: Downcode filtered alleles to reference, combine PLs
+          using minimum for each overloaded genotype, and shift so the
+          overall minimum PL is 0.
+        - GQ: The second-lowest PL (after shifting).
+        """
+        ds = self._ds
+        newPL = functions.cond(
+            functions.is_defined(ds.PL),
+            functions.range(0, self.new_v.num_genotypes()).map(
+                lambda newi: (functions.range(0, ds.v.num_genotypes()).filter(
+                    lambda oldi: bind(
+                        functions.call(newi),
+                        lambda oldc: functions.gt_index(self.old_to_new[oldc.gtj()],
+                                                        self.old_to_new[oldc.gtk()]) == newi))
+                              .map(lambda oldi: ds.PL[oldi]).min())),
```

Comment:
would be easier to read with .filter and .map similarly indented, see comment below